### ISSUE TITLE
Invalid character can ruin a .leo file

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#1510]
A `0x01` character in the .leo file produces the traceback given below.
```
bad .leo file: WORK.leo
g.toUnicode: unexpected argument of type ParseError
toUnicode openLeoFile,getLeoFile,readFile,readWithElementTree


Traceback (most recent call last):

  File "E:\git\leo\leo\core\leoCommands.py", line 2278, in executeAnyCommand
    return command(event)

  File "E:\git\leo\leo\core\leoGlobals.py", line 240, in commander_command_wrapper
    method(event=event)

  File "E:\git\leo\leo\commands\commanderFileCommands.py", line 261, in open_outline
    open_completer(c, closeFlag, fileName)

  File "E:\git\leo\leo\commands\commanderFileCommands.py", line 204, in open_completer
    c2 = g.openWithFileName(fileName, old_c=c)

  File "E:\git\leo\leo\core\leoGlobals.py", line 3920, in openWithFileName
    return g.app.loadManager.loadLocalFile(fileName, gui, old_c)

  File "E:\git\leo\leo\core\leoApp.py", line 3288, in loadLocalFile
    previousSettings = lm.getPreviousSettings(fn)

  File "E:\git\leo\leo\core\leoApp.py", line 2241, in getPreviousSettings
    c = lm.openSettingsFile(fn)

  File "E:\git\leo\leo\core\leoApp.py", line 2412, in openSettingsFile
    ok = c.fileCommands.openLeoFile(theFile, fn,

  File "E:\git\leo\leo\core\leoFileCommands.py", line 692, in openLeoFile
    ok, ratio = self.getLeoFile(

  File "E:\git\leo\leo\core\leoFileCommands.py", line 559, in getLeoFile
    v = FastRead(c, self.gnxDict).readFile(fileName)

  File "E:\git\leo\leo\core\leoFileCommands.py", line 60, in readFile
    v, g_element = self.readWithElementTree(path, s)

TypeError: cannot unpack non-iterable NoneType object

Traceback (most recent call last):
  File "E:\git\leo\leo\plugins\qt_frame.py", line 1638, in closeEvent
    g.app.gui.close_event(event)
AttributeError: 'NullGui' object has no attribute 'close_event'
```

**Comment by tbpassin** [comment]
"A 0x01 character in the .leo file produces the traceback given below. There are two possible fixes:

    Prohibit bad characters when writing .leo files.
    Recover from xml read errors."

This is a tough problem.  Yes, bad characters shouldn't be written to .leo files.  The question is: what to do if a bad character is found before the file is written.

Recovering from an xml error has the same error.  Yes, recovery could be possible, but what should be done with the xml element that contains that character?  If it were omitted, that could cause serious problems in the loaded file.

My first thought is to run a diagnostic method on the ,leo file when this condition is discovered, to notify the user as to the line and character position as closely as possible, and to suggest that the user should edit the file in a text editor to fix it.  The actual line or lines in question should be displayed so the user knows what to look for.  That's not wonderful, but I can't see how an automatic solution is going to be possible:

There are going to be basically four ways this could happen:

1. There is a non-utf-8 byte or bytes in the file (I know that other encodings are possible with XML, but I'm pretty sure that Leo won't generate non-utf-8 files).

2. The file could contain only legal utf-8 byte sequences, but not be well-formed for some unknown reason.  Well-formedness can be checked quickly.

3. The file could be well-formed, but some piece of data that Leo needs to function has been corrupted.

4. There could be a utf-8-legal character that isn't allowed by XML.  That seems to be the case here.

**Comment by edreamleo** [comment]
@tbpassin I agree that this is a non-trivial problem.

**Comment by edreamleo** [comment]
Fixed at 1f2e2be in xml.

**Comment by edreamleo** [comment]
I am going to reopen this issue.  See this post:
https://groups.google.com/g/leo-editor/c/m1cIciVq_6g/m/9Zb0aH3ACAAJ

**Comment by edreamleo** [comment]
Fixed at rev 5234c8b in devel. The test failure noted above results from the failure to create a Python 3.6 environment.
